_posts/2023-08-14-clear-specification-and-implementation.markdown

```diff
@@ -0,0 +1,610 @@
+
+    // The actual content of the stack is stored in a linked sequence of
+    // Node objects, defined below.
+    var top : Node?<T>
+
+    // The model is exactly the same as for the array-based
+    // implementation.
+    ghost var model : StackModel<T>
+
+    // The footprint keeps track of all objects reachable from this
+    // object. This is used to prove termination.
+    ghost var footprint : set<object>
+
+    // Validity is a bit more complex, as it needs to recursively
+    // traverse the list and use the associated footprint to ensure
+    // termination.
+    ghost predicate ValidModel()
+      reads this, top, footprint
+    {
+      this in footprint &&
+      model == NodeModel(top) &&
+      (top != null ==>
+        top in footprint &&
+        this !in top.footprint &&
+        top.footprint <= footprint &&
+        top.ValidModel())
+    }
+
+    // As before, an initial stack is easy to construct and show to be
+    // empty and valid.
+    constructor InitStack()
+      ensures ValidModel()
+      ensures IsEmpty()
+    {
+      top := null;
+      model := EmptyModel();
+      footprint := {this};
+    }
+
+    // Pushing a value is simpler than in the array case, and satisfies
+    // the same equivalence with the specification.
+    method Push(val : T)
+      requires ValidModel()
+      modifies this
+      ensures ValidModel()
+      ensures !IsEmpty()
+      ensures model == PushModel(old(model), val)
+    {
+      top := new Node<T>.InitNode(val, top);
+      footprint := footprint + top.footprint;
+      model := PushModel(model, val);
+    }
+
+    // Popping a value is very symmetrical with pushing.
+    method Pop() returns (result : T)
+      requires ValidModel()
+      requires !IsEmpty()
+      modifies this
+      ensures ValidModel()
+      ensures model == PopModel(old(model)).1
+      ensures result == PopModel(old(model)).0
+    {
+      result := top.val;
+      top := top.next;
+      model := PopModel(model).1;
+    }
+
+    // Checking emptiness is very simple, as in the array case.
+    predicate IsEmpty()
+      reads this, top, footprint
+      requires ValidModel()
+      ensures IsEmpty() <==> IsEmptyModel(model)
+    {
+      top == null
+    }
+  }
+
+  // This class, that implements the actual linked data structure, is
+  // where the complexity lives. If you already had a linked list
+  // implementation, you could potentially use that, and share the
+  // burden of implementation and verification with any other uses of
+  // linked lists. Although it would require some cleverness to
+  // integrate the ghost model and node validity notions with a
+  // general-purpose list.
+  class Node<T> {
+
+    // Each node has a value.
+    var val : T
+
+    // Each node points to the next node, which may be null if this is
+    // the last node (the bottom of the stack).
+    var next : Node?<T>
+
+    // The model is the same as in the array-based case.
+    ghost var model : StackModel<T>
```

Comment:
Maybe a comment about the model being a simpler but perhaps slower implementation, but no impact on performance because of ghost?

Reply:
Good idea.